### PR TITLE
feat: last mile approach to speed up contract read

### PR DESCRIPTION
**Pull request by johnletey**
This PR adds local state caching to improve contract read times.

### CODE REVIEW

File: src/contract-read.ts [hunk 1]
```diff
@@ -28,6 +34,14 @@ export async function readContract(
     height = networkInfo.height;
   }
 
+  if (contractId in cache) {
+    if (height in cache[contractId]) {
+      const res = JSON.parse(cache[contractId][height]);
+
+      return returnValidity ? { state: res.state, validity: res.validity } : res.state;
```
Comment (hlolli) [inline src/contract-read.ts:41]: perhaps this would require unrelated changes. But returning both object of state and validity, or just state, creates an annoying typing situation. Could it be a return of `{ state: IState, validity?: IValidity }`, then validity can either be undefined or defined.

File: src/contract-read.ts [hunk 2]
```diff
@@ -28,6 +34,14 @@ export async function readContract(
     height = networkInfo.height;
   }
 
+  if (contractId in cache) {
+    if (height in cache[contractId]) {
+      const res = JSON.parse(cache[contractId][height]);
```
Comment (hlolli) [inline src/contract-read.ts:39]: Two things here, the res would benefit from being typed, as well as wrap the json.parse into a try/catch block, in which case we may need to think how to best return errors. Like undefined state and validity as { validity: { error: "error while parsing json" } }, just thinking out loud here.

File: src/contract-read.ts [hunk 3]
```diff
@@ -7,6 +7,12 @@ import GQLResultInterface, { GQLEdgeInterface, GQLTransactionsResultInterface }
 
 import SmartWeaveError, { SmartWeaveErrorType } from './errors';
 
+const cache: {
```
Comment (ppedziwiatr) [inline src/contract-read.ts:10]: how long does this cache live in the memory? I'm asking because of current issues with memory leaks ( https://discord.com/channels/357957786904166400/756557551234973696/857784751149940736 ) - I wonder if it doesn't introduce the risk of another leak...

Comment (johnletey) [reply src/contract-read.ts:10]: @ppedziwiatr The cache lives in memory until the function is done running.

Comment (ppedziwiatr) [reply src/contract-read.ts:10]: As discussed here: https://discord.com/channels/357957786904166400/756557551234973696/860136978380095509 - it is a globally scoped cache and lives in the memory after this function exits - just leaving for future reference.
